### ISSUE TITLE
%s/foo//gn doesn't work

### ISSUE THREAD

**Issue by daphnisz** [emacs-evil/evil#811]
In vim, you can type `:%s/foo//gn` to get the number of occurrences of `foo` without changing anything, but this doesn't work in evil.

**Comment by edkolev** [comment]
I tried adding support for `n` in `evil-ex-substitute`, but it turned out a big task:

It isn't a simple task when both `g` and `n` are given, because with `g` all the replacing is done with `perform-replace`. Hence, `evil-ex-substitute` will have to be refactored to not use `perform-replace` and that is a change I'm not comfortable doing. The diff which add support for `n` (but *doesn't* work with both `gn`) is pasted below in case some wants to pick that up.

I would like to suggest this alternative approach: instead of supporting `n`, we can show the number of matches like so:

![screen shot 2017-04-19 at 09 41 32](https://cloud.githubusercontent.com/assets/1532071/25166528/7f8b477a-24e4-11e7-8417-938bae572c50.png)

The code is very simple, it simply prints the count of overlays (no extra counting is done).
We may add a configuration variable which, if nil, will not bother the user with the number of matches.
I can wrap this in a PR if the maintainers think this this is a good change.

The half-working diff adding support for `n`:

```diff
git diff evil-commands.el
diff --git a/evil-commands.el b/evil-commands.el
index 44ec3aa..281f904 100644
--- a/evil-commands.el
+++ b/evil-commands.el
@@ -3329,7 +3329,8 @@ resp.  after executing the command."
   (setq replacement (or replacement ""))
   (setq evil-ex-last-was-search nil)
   (let* ((flags (append flags nil))
-         (confirm (memq ?c flags))
+         (count-only (memq ?n flags))
+         (confirm (and (not count-only) (memq ?c flags))) ;; ignore "c" flag if "n" if given
          (case-fold-search (evil-ex-pattern-ignore-case pattern))
          (case-replace case-fold-search)
          (evil-ex-substitute-regex (evil-ex-pattern-regex pattern)))
@@ -3420,14 +3421,16 @@ resp.  after executing the command."
                                      (line-end-position) t nil)
               (setq evil-ex-substitute-nreplaced
                     (1+ evil-ex-substitute-nreplaced))
-              (evil-replace-match evil-ex-substitute-replacement
-                                  (not case-replace))
+              (unless count-only
+                (evil-replace-match evil-ex-substitute-replacement
+                                    (not case-replace)))
               (setq evil-ex-substitute-last-point (point)))))

         (while markers (move-marker (pop markers) nil))
         (goto-char evil-ex-substitute-last-point)

-        (message "Replaced %d occurrence%s"
+        (message "%s %d occurrence%s"
+                 (if count-only "Found" "Replaced")
                  evil-ex-substitute-nreplaced
                  (if (/= evil-ex-substitute-nreplaced 1) "s" ""))))
     (evil-first-non-blank)))
```

**Comment by wasamasa** [comment]
This is an interesting way of solving it, not entirely sure on it until I've seen the fully working diff. @TheBB Thoughts?

**Comment by edkolev** [comment]
Sorry, I misled you - the count of overlays is the count of *visible* matches, not total matches. I'll write here if I manage to get something working.

**Comment by wasamasa** [comment]
This has been solved thanks to #832!
